src/puppetlabs/puppetdb/queue.clj

```diff
@@ -30,16 +33,79 @@
+      (re-pattern (str "("
+                       (->> forbidden-chars
+                           (map #(format "\\Q%s\\E" %)) ; escape regex chars
+                           (str/join "|"))
+                       ")"))
+      "-")))
+
+(defn max-certname-length
+  "Given the received-at time, command and command version for a metadata
+  string, returns the maximum allowable length (in bytes) of a certname in that
+  string. Note that this length is only achievable if the certname does not
+  need to be sanitized, since sanitized certnames always have a SHA1 hash
+  appended."
+  [received command version]
+  (let [time-length (-> received tcoerce/to-long str utf8-length)
+        command-length (utf8-length command)
+        version-length (utf8-length (str version))
+        field-separators 3]
+    (- 255 ; overall filename length limit
+       time-length command-length version-length
+       3 ; number of field separators (underscores)
+       (utf8-length ".json"))))
+
+(defn truncated-certname-length
+  "Given the received-at time, command, and command version that will be in
+  a metadata string, returns the length (in bytes) to truncate certnames to when
+  they are longer than the `max-certname-length` for the string or they contain
+  characters that must be sanitized. This is less than the string's
+  `max-certname-length` to leave space for the SHA1 hash."
+  [received command version]
+  (- (max-certname-length received command version)
+     1 ; for the additional field separator between certname and hash
+     (utf8-length (kitchensink/utf8-string->sha1 ""))))
+
+(defn embeddable-certname
+  "Takes all the components of a metadata string, and returns a version of the
+  certname that's safe to embed in the metadata string. It will be sanitized to
+  replace any illegal filesystem characters and underscores with dashes, and
+  will be truncated if the original version will cause the metadata string to
+  exceed 255 characters."
+  [received command version certname]
+  (let [cn-length (utf8-length certname)
+        trunc-length (truncated-certname-length received command version)
+        sanitized-certname (sanitize-certname certname)]
+    (if (or (> cn-length (max-certname-length received command version))
+            (and (not= certname sanitized-certname)
+                 (> cn-length trunc-length)))
+      (utf8-truncate sanitized-certname trunc-length)
+      sanitized-certname)))
+
 (defn metadata-str [received command version certname]
-  (format "%d_%s_%d_%s.json"
-          (tcoerce/to-long received) command version certname))
+  (let [certname (or certname "unknown-host")
+        recvd-long (tcoerce/to-long received)
+        safe-certname (embeddable-certname received command version certname)]
+    (if (= safe-certname certname)
+      (format "%d_%s_%d_%s.json" recvd-long command version safe-certname)
+      (let [name-hash (kitchensink/utf8-string->sha1 certname)]
+        (format "%d_%s_%d_%s_%s.json"
```

Comment:
One issue that I think we should discuss. This PR basically changes a certname if that certname is too long before storing the command in stockpile. That metadata string is not what we're storing in memory, we're storing the pieces that form it. Let's say we have a catalog with a certname that is too long. We would encode it as described, however the "too long" certname would be in memory as the full string. PDB is stopped and started, we receive a new catalog from that same certname, we wouldn't recognize that those two nodes are the same. The one we read from disk would be represented in memory as "<way too...>-<some hash>" and the new one would be represented as "<way too long>".

Although I don't think the above is a big deal, more of a performance optimization related to our "bash in place" code, but would be curious how that complicates things. What got me thinking about this is interning certnames so save memory on enqueued commands.